File: testDeepConvolutionalEmbeddedClustering.py
from os import path
import antspynet
import keras.backend as K
import keras
import numpy as np

from keras.datasets import fashion_mnist

(x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()
number_of_clusters = 10

number_of_training_data = len(y_train)
number_of_testing_data = len(y_test)

x = x_train / 255
y = y_train

pretrain_epochs = 300
pretrain_batch_size = 256

# Instantiate the DEC model

dcec_model = antspynet.DeepEmbeddedClusteringModel(
   number_of_units_per_layer=(32, 64, 128, 10),
   number_of_clusters = number_of_clusters,
   convolutional = True, input_image_size = (28, 28, 1))

model_weights_file = "dcecAutoencoderModelWeights.h5"
if not path.exists(model_weights_file):
    dcec_model.pretrain(x=x, optimizer='adam', epochs=pretrain_epochs,
      batch_size=pretrain_batch_size)
    dcec_model.autoencoder.save_weights(model_weights_file)  
else:
    dcec_model.autoencoder.load_weights(model_weights_file)  

dcec_model.compile(optimizer=keras.optimizers.adam(), loss=('kld', 'mse'), loss_weights=(0.1, 1))

y_predicted = dcec_model.fit(x, max_number_of_iterations=20000, batch_size=256,
  tolerance=0.0001, update_interval=10)

File: testDeepConvolutionalEmbeddedClustering.R
library( keras )
library( ANTsRNet )

keras::backend()$clear_session()

# Let's use the fmnist data set.

fmnist <- dataset_fashion_mnist()

numberOfTrainingData <- length( fmnist$train$y )
numberOfTestingData <- length( fmnist$test$y )

inputImageSize <- dim( fmnist$test$x[1,,] )

x <- array( data = fmnist$train$x / 255, dim = c( numberOfTrainingData, inputImageSize, 1 ) )
y <- fmnist$train$y

numberOfClusters <- length( unique( fmnist$train$y ) )

pretrainEpochs <- 300L
pretrainBatchSize <- 256L

pretrainOptimizer <- optimizer_sgd( lr = 1.0, momentum = 0.9 )

# Instantiate the DCEC model

dcecModel <- DeepEmbeddedClusteringModel$new(
   numberOfUnitsPerLayer = c( 32, 64, 128, 10 ),
   numberOfClusters = numberOfClusters,
   convolutional = TRUE, inputImageSize = c( inputImageSize, 1 ) )

modelWeightsFile <- "dcecAutoencoderModelWeights.h5"
if( ! file.exists( modelWeightsFile ) )
  {
  dcecModel$pretrain( x = x, optimizer = 'adam',
    epochs = pretrainEpochs, batchSize = pretrainBatchSize )
  save_model_weights_hdf5( dcecModel$autoencoder, modelWeightsFile )
  } else {
  load_model_weights_hdf5( dcecModel$autoencoder, modelWeightsFile )
  }

dcecModel$compile( optimizer = 'adam', loss = c( 'kld', 'mse' ), loss_weights = c( 0.1, 1 ) )

yPredicted <- dcecModel$fit( x, maxNumberOfIterations = 2e4, batchSize = 256,
  tolerance = 1e-3, updateInterval = 10 )

